src/hardware/pcspeaker_discrete.cpp

```diff
@@ -0,0 +1,491 @@
+ *  with this program; if not, write to the Free Software Foundation, Inc.,
+ *  51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+ */
+
+#include "pcspeaker_discrete.h"
+
+#include <algorithm>
+#include <array>
+#include <cmath>
+#include <limits>
+
+#include "checks.h"
+
+CHECK_NARROWING();
+
+// Determines if the inbound wave is square by inspecting
+// current and previous states.
+bool PcSpeakerDiscrete::IsWaveSquare()
+{
+	// When compared across time, this value describe an active PIT-state
+	// being toggled
+	constexpr auto was_toggled     = 0b10 + 0b11;
+	constexpr auto was_steadily_on = 0b11;
+	constexpr auto was_toggled_on  = 0b01;
+
+	// The sum of the previous port_b_state with the current port_b_state
+	// becomes a temporal-state
+	const auto temporal_pit_state = static_cast<int>(prev_pit_mode) + static_cast<int>(pit_mode);
+	const auto temporal_pwm_state = static_cast<int>(prev_port_b_state.timerGateAndSpeakerOutput) +
+	                                static_cast<int>(port_b_state.timerGateAndSpeakerOutput);
+
+	// We have a sine-wave if the PIT was steadily off and ...
+	if (temporal_pit_state == 0)
+		// The PWM toggled an ongoing PIT state or turned on PIT-mode
+		// from an off-state
+		if (temporal_pwm_state == was_toggled || temporal_pwm_state == was_steadily_on)
+			return false;
+
+	// We have a sine-wave if the PIT was steadily on and ...
+	if (temporal_pit_state == was_steadily_on)
+		// the PWM was turned on from an off-state
+		if (temporal_pwm_state == was_toggled_on)
+			return false;
+
+	return true;
+}
+
+void PcSpeakerDiscrete::AddDelayEntry(double index, double vol)
+{
+	if (IsWaveSquare()) {
+		vol *= sqw_scalar;
+// #define DEBUG_SQUARE_WAVE 1
+#ifdef DEBUG_SQUARE_WAVE
+		LOG_MSG("PCSPEAKER: square-wave [prev_pos=%u, prev_port_b_state=%u, port_b_state=%u, prev_pit=%lu, pit=%lu]",
+		        prev_pos,
+		        prev_port_b_state,
+		        port_b_state,
+		        prev_pit_mode,
+		        pit_mode);
+	} else {
+		LOG_MSG("PCSPEAKER: sine-wave [prev_pos=%u, prev_port_b_state=%u, port_b_state=%u, prev_pit=%lu, pit=%lu], ",
+		        prev_pos,
+		        prev_port_b_state,
+		        port_b_state,
+		        prev_pit_mode,
+		        pit_mode);
+#endif
+	}
+	entries.at(entries_queued) = {index, vol};
+	entries_queued++;
+
+#if 0
+	// This is extremely verbose; pipe the output to a file.
+	// Display the previous and current speaker modes w/ requested volume
+	if (fabs(vol) > amp_neutral)
+		LOG_MSG("PCSPEAKER: Adding pos=%3s, pit=%" PRIuPTR "|%" PRIuPTR
+		        ", pwm=%d|%d, volume=%6.0f",
+		        prev_pos > 0 ? "yes" : "no", prev_pit_mode,
+		        pit_mode, prev_port_b_state, port_b_state,
+		        static_cast<double>(vol));
+#endif
+}
+
+void PcSpeakerDiscrete::ForwardPIT(double newindex)
+{
+	auto passed     = (newindex - last_index);
+	auto delay_base = last_index;
+	last_index      = newindex;
+	switch (pit_mode) {
+	case PitMode::InterruptOnTC: return;
```

Comment:
What's "TC"?

Reply:
:+1: Expanded to Terminal Count (terminal count 

Good explanation here (which I've also included in the struct comment):

https://wiki.osdev.org/Programmable_Interval_Timer#Mode_0_.E2.80.93_Interrupt_On_Terminal_Count

Terminal Count is also used to indicate when a DMA transfer is completed "reached terminal count".  https://docs.freebsd.org/doc/3.3-RELEASE/usr/share/doc/handbook/dma.html